runtime/near-vm-logic/src/logic.rs

```diff
@@ -943,14 +944,86 @@ impl<'a> VMLogic<'a> {
     pub fn keccak512(&mut self, value_len: u64, value_ptr: u64, register_id: u64) -> Result<()> {
         self.gas_counter.pay_base(keccak512_base)?;
         let value = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
-        self.gas_counter.pay_per_byte(keccak512_byte, value.len() as u64)?;
+        self.gas_counter.pay_per(keccak512_byte, value.len() as u64)?;
 
         use sha3::Digest;
 
         let value_hash = sha3::Keccak512::digest(&value);
         self.internal_write_register(register_id, value_hash.as_slice().to_vec())
     }
 
+    /// Hashes the given value using RIPEMD-160 and returns it into `register_id`.
+    ///
+    /// # Errors
+    ///
+    /// If `value_len + value_ptr` points outside the memory or the registers use more memory than
+    /// the limit with `MemoryAccessViolation`.
+    ///
+    /// # Cost
+    ///
+    ///  Where `message_blocks` is `max(message_len / 8, 1)`.
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes + ripemd160_base + ripemd160_block * message_blocks`
+    pub fn ripemd160(&mut self, value_len: u64, value_ptr: u64, register_id: u64) -> Result<()> {
+        use num_integer::Integer;
+
+        self.gas_counter.pay_base(ripemd160_base)?;
+        let value = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
+
+        let message_blocks = (value_len + 9).div_ceil(&64);
+
+        self.gas_counter.pay_per(ripemd160_block, message_blocks)?;
+
+        use ripemd160::Digest;
+
+        let value_hash = ripemd160::Ripemd160::digest(&value);
+        self.internal_write_register(register_id, value_hash.as_slice().to_vec())
+    }
+
+    /// Recovers an ECDSA signer address and returns it into `register_id`.
+    ///
+    /// Takes in an additional flag to check for malleability of the signature
+    /// which is generally only ideal for transactions.
+    ///
+    /// Returns a bool indicating success or failure as a `u64`.
+    ///
+    /// # Errors
+    ///
+    /// * If `hash_ptr`, `r_ptr`, or `s_ptr` point outside the memory or the registers use more
+    ///   memory than the limit, then returns `MemoryAccessViolation`.
+    ///
+    /// # Cost
+    ///
+    /// `base + write_register_base + write_register_byte * 20 + ecrecover_base`
+    pub fn ecrecover(
+        &mut self,
+        hash_ptr: u64,
+        sig_ptr: u64,
+        malleability_flag: u8,
+        register_id: u64,
+    ) -> Result<u64> {
+        self.gas_counter.pay_base(ecrecover_base)?;
+
+        let mut signature_bytes = [0u8; 65];
+        self.memory_get_into(sig_ptr, &mut signature_bytes)?;
+        let signature = Secp256K1Signature::from(signature_bytes);
+
+        if malleability_flag == 1 {
+            if !signature.check_signature_values() {
+                return Ok(false as u64);
+            }
+        }
+
+        let mut hash_bytes = [0u8; 32];
+        self.memory_get_into(hash_ptr, &mut hash_bytes)?;
```

Comment:
Let's move this reading form memory *before* the if where we check `mallability_flag`. As written now, the code has weird behavior that, when it passes an out of bounds `hash_ptr`, *but* also passes invalid signature, it actually works. It seems better if invalid memory accesses are always caught.

Reply:
Yep, was late when I added that in. Good catch. Fixed: https://github.com/near/nearcore/pull/4380/commits/0acaa6700500576bd116b6fb8bfa6c0b47833838